### PR TITLE
Add support for Cirun on self-hosted GHA runners

### PR DESCRIPTION
**Pull request by jaimergp**
Checklist
* [x] Added a ``news`` entry

* Part of https://github.com/conda-forge/conda-forge.github.io/issues/1272.
* Testing at:
  * https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/446
  * https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/466

### CODE REVIEW

File: conda_smithy/cirun_utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import os
+
+from cirun import Cirun
+
+CIRUN_INSTALLATION_ID = os.environ.get("CIRUN_INSTALLATION_ID", 18453316)
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:5]: Add a comment about how to obtain this magic number, and how this is conda-forge's.

File: conda_smithy/cirun_utils.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+def remove_repo_from_cirun_resource(owner, repo, resource):
+    """Revoke access to a cirun resource to a particular repository, with a particular policy."""
+    cirun = _get_cirun_client()
+    print(f"Removing repo {owner}/{repo} from resource {resource}.")
+    response = cirun.remove_repo_from_resources(owner, repo, [resource])
+    print(f"response: {response} | {response.json().keys()}")
+    return response
+
+
+def _get_cirun_client():
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:48]: Is it worth caching this or do we actually want a separate instance each time?

File: conda_smithy/cirun_utils.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    return cirun.set_repo(
+        f"{owner}/{repo}", installation_id=CIRUN_INSTALLATION_ID
+    )
+
+
+def add_repo_to_cirun_resource(owner, repo, resource, cirun_policy_args):
+    """Grant access to a cirun resource to a particular repository, with a particular policy."""
+    cirun = _get_cirun_client()
+    policy_args = {"pull_request": False}
+    if cirun_policy_args:
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:22]: What's the expected type/schema of `cirun_policy_args`? A list of str? A mapping?

File: conda_smithy/cirun_utils.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    assert CIRUN_INSTALLATION_ID
+    return cirun.set_repo(
+        f"{owner}/{repo}", installation_id=CIRUN_INSTALLATION_ID
+    )
+
+
+def add_repo_to_cirun_resource(owner, repo, resource, cirun_policy_args):
+    """Grant access to a cirun resource to a particular repository, with a particular policy."""
+    cirun = _get_cirun_client()
+    policy_args = {"pull_request": False}
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:21]: Should we hardcode the defaults in this helper or should that be something done in the Cirun client (or backend)? I am a bit concerned we need to release a new conda-smithy if we want to change this default and it doesn't look like it's the best place to own that decision. Maybe the `.cirun` global yml?

File: conda_smithy/cirun_utils.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    print(
+        f"Adding repo {owner}/{repo} to resource {resource} with policy_args: {policy_args}"
+    )
+    response = cirun.add_repo_to_resources(
+        owner,
+        repo,
+        resources=[resource],
+        teams=[repo],
+        policy_args=policy_args,
+    )
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:34]: Is the client checking for bad http responses or do we need to `response.raise_for_status()` here?

File: conda_smithy/cli.py [hunk 6]
```diff
@@ -208,7 +210,7 @@ def __init__(self, parser):
         # conda-smithy register-ci ./
         super(RegisterCI, self).__init__(
             parser,
-            "Register a feedstock at the CI " "services which do the builds.",
+            "Register a feedstock at the CI services which do the builds.",
```
Comment (jaimergp) [inline conda_smithy/cli.py:213]: Thanks :D

File: conda_smithy/cli.py [hunk 7]
```diff
@@ -257,6 +260,22 @@ def __init__(self, parser):
             action="append",
             help="drone server URL to register this repo. multiple values allowed",
         )
+        scp.add_argument(
+            "--cirun-resources",
+            action="append",
+            help="cirun resources to enable for this repo. multiple values allowed",
+        )
+        scp.add_argument(
+            "--cirun-policy-args",
```
Comment (jaimergp) [inline conda_smithy/cli.py:269]: Where can we find a list of possible policy args?

File: conda_smithy/cli.py [hunk 8]
```diff
@@ -257,6 +260,22 @@ def __init__(self, parser):
...
+            action="append",
+            help="extra arguments for cirun policy to create for this repo. multiple values allowed",
+        )
+        scp.add_argument(
+            "--remove",
+            action="store_true",
+            help="Revoke access to the configured CI services. "
+            "Only available for Travis and Cirun for now",
```
Comment (jaimergp) [inline conda_smithy/cli.py:277]: Is it really available for Travis? 🤔

File: requirements.txt [hunk 9]
```diff
@@ -16,3 +16,4 @@ shellcheck
 scrypt
 license-expression
 setuptools_scm >=7
+cirun
```
Comment (jaimergp) [inline requirements.txt:19]: In https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/446 we need to install cirun from `main`. Is that the case here too?

**Review by jaimergp — COMMENTED** [review]
Thanks Amit, this looks very well put together and integrates well with https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/446. I only have some questions about some hardcoded defaults and small nits otherwise.

File: conda_smithy/cli.py [hunk 10]
```diff
@@ -257,6 +260,22 @@ def __init__(self, parser):
             action="append",
             help="drone server URL to register this repo. multiple values allowed",
         )
+        scp.add_argument(
+            "--cirun-resources",
```
Comment (jaimergp) [inline conda_smithy/cli.py:264]: Is this required when `cirun` is enabled? If optional, we need to set the default value to an iterable (not `None`), because otherwise we hit this bug:

```
conda-smithy register-ci --feedstock_directory . --organization conda-forge --without-azure --without-travis --without-circle --without-appveyor --without-drone --without-webservice --without-anaconda-token 
No circle token.  Create a token at https://circleci.com/account/api and
put it in ~/.conda-smithy/circle.token
No appveyor token. Create a token at https://ci.appveyor.com/api-token and
Put one in ~/.conda-smithy/appveyor.token
No drone token. Create a token at https://cloud.drone.io/account and
Put one in ~/.conda-smithy/drone.token
No anaconda token. Create a token via
  anaconda auth --create --name conda-smithy --scopes "repos conda api"
and put it in ~/.conda-smithy/anaconda.token
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.22
Adding in variants from internal_defaults
Adding in variants from /Users/jrodriguez/devel/cf-autotick-bot-test-package-feedstock/recipe/conda_build_config.yaml
CI Summary for conda-forge/cupy-feedstock (can take ~30s):
Warning: By not registering an Anaconda/Binstar tokenyour feedstock CI may not be able to upload packagesto anaconda.org by default. It is recommended to set`upload_packages: False` per provider field inconda-forge.yml to disable package uploads.
Travis registration disabled.
Circle registration disabled.
Azure registration disabled.
Appveyor registration disabled.
Drone registration disabled.
Traceback (most recent call last):
  File "/Users/jrodriguez/.local/anaconda/envs/smithy-dev/bin/conda-smithy", line 33, in <module>
    sys.exit(load_entry_point('conda-smithy', 'console_scripts', 'conda-smithy')())
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/cli.py", line 712, in main
    args.subcommand_func(args)
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/cli.py", line 386, in __call__
    for resource in args.cirun_resources:
TypeError: 'NoneType' object is not iterable
```

If required, then we need to error out properly.

File: conda_smithy/cli.py [hunk 11]
```diff
@@ -257,6 +260,22 @@ def __init__(self, parser):
...
+            action="append",
+            help="cirun resources to enable for this repo. multiple values allowed",
+        )
+        scp.add_argument(
+            "--cirun-policy-args",
+            action="append",
```
Comment (jaimergp) [inline conda_smithy/cli.py:270]: Same comment about `--cirun-resources` and `None` as default. We need some extra logic or config here to ensure it's properly handled.

File: conda_smithy/cli.py [hunk 12]
```diff
@@ -345,6 +376,31 @@ def __call__(self, args):
...
+                for resource in to_remove:
+                    conda_smithy.cirun_utils.remove_repo_from_cirun_resource(
+                        owner, repo, resource
+                    )
+            else:
+                print(
+                    f"Cirun Registration: resources to add to: {owner}/{repo}"
+                )
+                conda_smithy.cirun_utils.enable_cirun_for_project(owner, repo)
+                for resource in args.cirun_resources:
```
Comment (jaimergp) [inline conda_smithy/cli.py:397]:
```suggestion
                for resource in (args.cirun_resources or ()):
```

Fix for the bug above.

File: conda_smithy/cirun_utils.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    print(f"response: {response} | {response.json().keys()}")
+    return response
+
+
+def _get_cirun_client():
+    try:
+        return Cirun()
+    except KeyError:
+        raise RuntimeError(
+            "You must have CIRUN_API_KEY defined to do Cirun CI registration"
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:53]:
```suggestion
            "You must have CIRUN_API_KEY defined to do Cirun CI registration. "
```

File: conda_smithy/cli.py [hunk 14]
```diff
@@ -345,6 +376,31 @@ def __call__(self, args):
...
+                print(f"Cirun Registration: resources to remove: {to_remove}")
+                for resource in to_remove:
+                    conda_smithy.cirun_utils.remove_repo_from_cirun_resource(
+                        owner, repo, resource
+                    )
+            else:
+                print(
+                    f"Cirun Registration: resources to add to: {owner}/{repo}"
+                )
+                conda_smithy.cirun_utils.enable_cirun_for_project(owner, repo)
```
Comment (jaimergp) [inline conda_smithy/cli.py:396]:
```suggestion
                response = conda_smithy.cirun_utils.enable_cirun_for_project(owner, repo)
                response.raise_for_status()
```

We do not want to continue on errors, do we? Maybe we should even do that on the utility functions.

Otherwise, with a bad token, we get:

```
Cirun Registration
Cirun Registration: resources to add to: conda-forge/cf-autotick-bot-test-package-feedstock
Enabling cirun for conda-forge/cf-autotick-bot-test-package-feedstock ...
──────────────────────────────────────────────────────  ──────────────────────────────────────────────────────
Error
{
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest"
}
[18:50:03] Response Status Code: 401                                                               utils.py:52
──────────────────────────────────────────────────────  ──────────────────────────────────────────────────────
Heroku webservice registration disabled.

CI services have been enabled. You may wish to regenerate the feedstock.
Any changes will need commiting to the repo.
```

But the request actually failed.

File: conda_smithy/cirun_utils.py (hunk 4 above)
Comment (aktech) [reply conda_smithy/cirun_utils.py:21]: >     policy_args = {"pull_request": False}

This is already set as default in cirun backend, I'll set the default to none here.

File: conda_smithy/cirun_utils.py (hunk 5 above)
Comment (aktech) [reply conda_smithy/cirun_utils.py:34]: Yes, client is check for bad response.

File: conda_smithy/cli.py (hunk 7 above)
Comment (aktech) [reply conda_smithy/cli.py:269]: At this point only `pull_request` is supported, but for future reference I'll add this in the client documentation.

File: conda_smithy/cli.py (hunk 8 above)
Comment (aktech) [reply conda_smithy/cli.py:277]: Nope, removed.

File: requirements.txt (hunk 9 above)
Comment (aktech) [reply requirements.txt:19]: latest version is already released, so this is fine to use

File: conda_smithy/cli.py (hunk 11 above)
Comment (aktech) [reply conda_smithy/cli.py:271]: this property can be None

File: conda_smithy/cli.py (hunk 7 above)
Comment (jaimergp) [reply conda_smithy/cli.py:270]: Ok, then let's add a link to the docs in the help arg when it's available.

File: conda_smithy/cli.py (hunk 14 above)
Comment (jaimergp) [reply conda_smithy/cli.py:397]: The cirun client will check for errors now.

File: conda_smithy/configure_feedstock.py [hunk 15]
```diff
@@ -1824,12 +1811,43 @@ def _load_forge_config(forge_dir, exclusive_config_file, forge_yml=None):
...
+                    "x64",
+                    "self-hosted",
+                    "cirun-openstack-cpu",
+                ],
+                "cloud": "openstack",
+                "instance_type": "ci_medium",
+                "machine_image": "ubuntu-2204-cloud-jammy-20221104",
+                "region": "RegionOne",
+            },
+        ],
```
Comment (jaimergp) [inline conda_smithy/configure_feedstock.py:1850]: This should go in `conda-forge/.cirun`.

File: conda_smithy/templates/github-actions.yml.tmpl [hunk 16]
```diff
@@ -38,66 +78,34 @@ jobs:
...
-            labels: ['macOS', 'self-hosted', 'ARM64']
+            labels:
+              - self-hosted
+              - {{ platform_vars[data.build_platform]['self_hosted_label_os'] }}
+              {%- if platform_vars[data.build_platform]['self_hosted_label_arch'] %}
+              - {{ platform_vars[data.build_platform]['self_hosted_label_arch'] }}
+              {%- endif %}
+              {%- for label in github_actions.self_hosted_labels %}
+                {%- if 'gpu' in label %}
+                  {%- set ns.gpu = True %}
```
Comment (jaimergp) [inline conda_smithy/templates/github-actions.yml.tmpl:93]: We are using `ns` as a global namespace to make sure that the line below for `CONDA_FORGE_DOCKER_RUN_ARGS` can see this flag. Otherwise, scoping rules will prevent it.

File: conda_smithy/templates/github-actions.yml.tmpl [hunk 17]
```diff
@@ -38,66 +78,34 @@ jobs:
...
+                {%- endif %}
+                {%- if label.startswith('cirun-') %}
+                  {%- set label = label ~ "--${{ github.run_id }}-" ~ data.short_config_name %}
```
Comment (jaimergp) [inline conda_smithy/templates/github-actions.yml.tmpl:96]: We need unique labels so the on-demand creation mechanism works without issues. See https://docs.cirun.io/reference/unique-runner-labels.

**Review by jaimergp — COMMENTED** [review]
LGTM. I won't approve yet because I want to run some tests locally but so far this round of review has been successfully addressed.

We need some documentation upstream (`cirun` client) and I need to check that #1761 didn't introduce conflicts here.

Thanks!

File: conda_smithy/cli.py (hunk 7 above)
Comment (aktech) [reply conda_smithy/cli.py:270]: add a link to cirun client docs at the top of `conda_smithy/cirun_utils.py`

File: conda_smithy/configure_feedstock.py (hunk 15 above)
Comment (aktech) [reply conda_smithy/configure_feedstock.py]: Removed and created a PR here: https://github.com/conda-forge/.cirun/pull/1

File: conda_smithy/templates/github-actions.yml.tmpl [hunk 18]
```diff
@@ -38,66 +86,34 @@ jobs:
...
+                {%- endif %}
+              - {{ label }}
+              {%- endfor %}
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda_smithy/templates/github-actions.yml.tmpl]: Done! Please see comments about deepcopies, but I think we are ok.

File: conda_smithy/templates/github-actions.yml.tmpl [hunk 19]
```diff
@@ -38,66 +39,16 @@ jobs:
...
-        done
-        sudo apt-get purge -y -f firefox \
-                                 google-chrome-stable \
-                                 microsoft-edge-stable
-        sudo apt-get autoremove -y >& /dev/null
-        sudo apt-get autoclean -y >& /dev/null
-        sudo docker image prune --all --force
-        df -h
-{%- endif %}
-
```
Comment (isuruf) [inline conda_smithy/templates/github-actions.yml.tmpl:100]: Bring this back?

File: conda_smithy/configure_feedstock.py [hunk 20]
```diff
@@ -1266,6 +1265,66 @@ def _github_actions_specific_setup(
...
+            "os": "ubuntu",
+            "self_hosted_labels": ("linux", "x64"),
+        },
+        "linux-aarch64": {
+            "os": "ubuntu",
+            "self_hosted_labels": ("linux", "ARM64"),
+        },
+        "linux-ppc64le": {
+            "os": "ubuntu",
+            "self_hosted_labels": ("linux", ""),
```
Comment (isuruf) [inline conda_smithy/configure_feedstock.py:1290]: I don't understand the empty label

File: conda_smithy/configure_feedstock.py [hunk 21]
```diff
@@ -1836,6 +1904,10 @@ def _load_forge_config(forge_dir, exclusive_config_file, forge_yml=None):
         },
         "github_actions": {
             "self_hosted": False,
+            "self_hosted_labels": [],
```
Comment (isuruf) [inline conda_smithy/configure_feedstock.py:1907]: Can this be done per build_platform? We would need different labels for different providers like cirun and win-arm64 workers.

Comment (jaimergp) [reply conda_smithy/configure_feedstock.py:1907]: Oh, yes, I see what you mean. This would also help if we need to use the GPUs only on Linux-64!

File: conda_smithy/configure_feedstock.py (hunk 20 above)
Comment (jaimergp) [reply conda_smithy/configure_feedstock.py:1290]: Oops, I left that as a placeholder because I wanted to see if I could find the PPC label on the github/runner repo but turns out it's not supported, so we need to remove it.

File: conda_smithy/templates/github-actions.yml.tmpl (hunk 19 above)
Comment (jaimergp) [reply conda_smithy/templates/github-actions.yml.tmpl:100]: Agh, my bad, sorry!

**Comment by jaimergp** [comment]
Hi @isuruf, I am using this PR to rerender [this one here (GPU tests for cupy)](https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/466) and it's erroring out with:


```
Traceback (most recent call last):
  File "/Users/jrodriguez/.local/anaconda/envs/smithy-dev/bin/conda-smithy", line 33, in <module>
    sys.exit(load_entry_point('conda-smithy', 'console_scripts', 'conda-smithy')())
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/cli.py", line 682, in main
    args.subcommand_func(args)
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/cli.py", line 542, in __call__
    self._call(args, tmpdir)
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/cli.py", line 547, in _call
    configure_feedstock.main(
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/configure_feedstock.py", line 2588, in main
    render_README(env, config, forge_dir, render_info)
  File "/Users/jrodriguez/devel/conda-smithy/conda_smithy/configure_feedstock.py", line 1713, in render_README
    channel_targets = list(dict.fromkeys(channel_targets))
TypeError: unhashable type: 'list'
```

`channel_targets` has this value `[['conda-forge', 'main'], ['conda-forge', 'main'], ['conda-forge', 'main'], ['conda-forge', 'main'], ['conda-forge', 'main'], ['conda-forge', 'main']]`

Is that a bug or do I need to change the conda-forge.yml / conda_build_config.yaml there?

**Comment by isuruf** [comment]
It's a bug from https://github.com/conda-forge/conda-smithy/pull/1752

File: conda_smithy/cirun_utils.py [hunk 22]
```diff
@@ -7,11 +7,25 @@
...
-CIRUN_INSTALLATION_ID = os.environ.get("CIRUN_INSTALLATION_ID", 18453316)
+
+@lru_cache
+def get_cirun_installation_id(owner: str) -> int:
+    # This ID needs a token with admin: org privileges.
+    # Hard-code instead for easier use
+    if owner == "conda-forge":
+        return 18453316
+    else:
+        from .github import gh_token, Github
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:18]: Any reason to have this import here instead of top-level?

File: conda_smithy/cirun_utils.py [hunk 23]
```diff
@@ -7,11 +7,25 @@
...
+        gh = Github(gh_token)
+        user = gh.get_user()
+        if (user.login == owner):
+            user_or_org = user
+        else:
+            user_or_org = gh.get_organization(owner)
+        for inst in user_or_org.get_installations:
+            if inst.raw_data["app_slug"] == "cirun-application":
+                return inst.app_id
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py]: Noice.

File: conda_smithy/cirun_utils.py [hunk 24]
```diff
@@ -7,11 +7,25 @@
...
+def get_cirun_installation_id(owner: str) -> int:
+    # This ID needs a token with admin: org privileges.
+    # Hard-code instead for easier use
+    if owner == "conda-forge":
+        return 18453316
+    else:
+        from .github import gh_token, Github
+        gh = Github(gh_token)
+        user = gh.get_user()
+        if (user.login == owner):
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py]: No brackets needed.

File: conda_smithy/cirun_utils.py [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def enable_cirun_for_project(owner: str, repo: str) -> Dict[str, Any]:
+    """Enable the cirun.io Github Application for a particular repository."""
+    print(f"Enabling cirun for {owner}/{repo} ...")
+    cirun = _get_cirun_client()
+    assert CIRUN_INSTALLATION_ID
+    return cirun.set_repo(
+        f"{owner}/{repo}", installation_id=CIRUN_INSTALLATION_ID
+    )
```
Comment (jaimergp) [inline conda_smithy/cirun_utils.py:39]:
```suggestion
    return cirun.set_repo(
        f"{owner}/{repo}", installation_id=get_cirun_installation_id()
    )
```

File: conda_smithy/cirun_utils.py (hunk 22 above)
Comment (isuruf) [reply conda_smithy/cirun_utils.py]: Moved to top level

**Comment by beckermr** [comment]
I don't think I have expertise to review this PR. However, can we make the docs PR sooner rather than later? I think that will help others figure out what to do with all of these new features.

**Comment by isuruf** [comment]
Docs PR: https://github.com/conda-forge/conda-forge.github.io/pull/2025

**Comment by isuruf** [comment]
@conda-forge/core, ready for a review

**Comment by jaimergp** [comment]
Basic docs have been merged. There's a chance we need to write more, but for now I think they are sufficient to get us unblocked here and merge to try things.

File: conda_smithy/cirun_utils.py [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+    """Revoke access to a cirun resource to a particular repository, with a particular policy."""
+    cirun = _get_cirun_client()
+    print(f"Removing repo {owner}/{repo} from resource {resource}.")
+    response = cirun.remove_repo_from_resources(owner, repo, [resource])
+    print(f"response: {response} | {response.json().keys()}")
+    return response
+
+
+@lru_cache
+def _get_cirun_client() -> Cirun:
```
Comment (xhochy) [inline conda_smithy/cirun_utils.py:75]: Moving the import down here would save us the `cirun` import if the code is not used. This doesn't need to be implemented to get it merged as we have been in general not optimizing the import times.
```suggestion
def _get_cirun_client() -> Cirun:
    from cirun import Cirun

```

File: conda_smithy/configure_feedstock.py [hunk 27]
```diff
@@ -2008,6 +2091,17 @@ def _load_forge_config(forge_dir, exclusive_config_file, forge_yml=None):
     if config["test"] is None:
         config["test"] = "all"
 
+    if config["github_actions"]["cancel_in_progress"] is None:
```
Comment (xhochy) [inline conda_smithy/configure_feedstock.py:2094]: What is the reasoning of this? Why bind `cancel_in_progress` to whether the runners are self-hosted?

Comment (jaimergp) [reply conda_smithy/configure_feedstock.py:2094]: To save CI resources by default. Self-hosted runners are usually going to be scarce or costly. If a user wants to not cancel-in-progress, they can explicitly say so, but by default it will cancel previous jobs when self-hosted runners are in use.

File: conda_smithy/cirun_utils.py (hunk 26 above)
Comment (isuruf) [reply conda_smithy/cirun_utils.py:75]: I don't think that would work as the return type needs the import.

Comment (xhochy) [reply conda_smithy/cirun_utils.py:75]: It would work with some `TYPE_CHECKING` trickery, but I don't think this is worth delaying this PR. Without measurements of the import times, I don't think it is worth further discussing it.

File: conda_smithy/configure_feedstock.py (hunk 27 above)
Comment (xhochy) [reply conda_smithy/configure_feedstock.py:2094]: I would also cancel hosted ones, too. They are scarce (only a limited pool of them) and can also be costly if you are not using your free minutes (or a runner that falls into the free minutes).

I would use the rather simple option of

```
if config["github_actions"]["cancel_in_progress"] is None:
    config["github_actions"]["cancel_in_progress"] = True
```

Probably you can just set the default to `False`?

Comment (isuruf) [reply conda_smithy/configure_feedstock.py:2094]: So, the cancel in progress here is configured to cancel all jobs in the same build. For eg: Linux builds will get cancelled if windows builds fail.

**Review by xhochy — APPROVED** [review]

File: conda_smithy/configure_feedstock.py [hunk 28]
```diff
@@ -695,7 +698,6 @@ def _render_ci_provider(
                 channel_target.startswith("conda-forge ")
                 and provider_name == "github_actions"
                 and not forge_config["github_actions"]["self_hosted"]
-                and os.path.basename(forge_dir) not in SERVICE_FEEDSTOCKS
```
Comment (beckermr) [inline conda_smithy/configure_feedstock.py:698]: @jaimergp why was this line removed?

Comment (beckermr) [reply conda_smithy/configure_feedstock.py:698]: I think we want it to be self-hosted or in service
